File: server/test/recipients.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { checkRecipient, isAllowedRecipient, REJECTION_MESSAGE } from '../src/recipients.js';

test('accepts addresses on the family domain', () => {
  const good = [
    'dad@refrigerationservices.com.au',
    'jane.doe@refrigerationservices.com.au',
    'service+jobs@refrigerationservices.com.au',
    'DAD@REFRIGERATIONSERVICES.COM.AU',
    '  spaced@refrigerationservices.com.au  ',
  ];
  for (const email of good) {
    assert.equal(isAllowedRecipient(email), true, `expected ${email} to be allowed`);
  }
});

test('rejects everything else, including lookalikes', () => {
  const bad = [
    'someone@gmail.com',
    'someone@refrigerationservices.com',            // missing .au
    'someone@refrigerationservices.com.au.evil.com', // suffix smuggling
    'evil.com@refrigerationservices.com.au.attacker.net',
    'someone@sub.refrigerationservices.com.au',      // subdomain is not the domain
    'someone@notrefrigerationservices.com.au',       // prefix smuggling
    'a@refrigerationservices.com.au, b@evil.com',    // list smuggled as one address
    'a@refrigerationservices.com.au;b@evil.com',
    '"a@refrigerationservices.com.au" <b@evil.com>',
    'a@refrigerationservices.com.au\nbcc: b@evil.com', // header injection attempt
    '@refrigerationservices.com.au',
    'refrigerationservices.com.au',
    '',
    null,
    undefined,
    42,
  ];
  for (const email of bad) {
    assert.equal(isAllowedRecipient(email), false, `expected ${JSON.stringify(email)} to be rejected`);
  }
});

test('checkRecipient returns the exact wording the business asked for', () => {
  assert.deepEqual(checkRecipient('stranger@gmail.com'), {
    ok: false,
    error: 'This email is not one of your family member',
  });
  assert.equal(REJECTION_MESSAGE, 'This email is not one of your family member');
});

test('checkRecipient distinguishes empty input from a disallowed domain', () => {
  const empty = checkRecipient('   ');
  assert.equal(empty.ok, false);
  assert.notEqual(empty.error, REJECTION_MESSAGE, 'blank input is a different problem to a wrong domain');
});

test('checkRecipient normalises accepted addresses', () => {
  assert.deepEqual(checkRecipient('  Dad@Refrigerationservices.Com.Au '), {
    ok: true,
    email: 'dad@refrigerationservices.com.au',
  });
});

File: android/app/src/test/java/au/com/refrigerationservices/weather/RecipientsTest.kt
package au.com.refrigerationservices.weather

import au.com.refrigerationservices.weather.data.Recipients
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Mirrors server/test/recipients.test.js. If these two ever disagree the phone
 * would either nag about valid addresses or wave through invalid ones.
 */
class RecipientsTest {

    @Test
    fun `accepts addresses on the family domain`() {
        listOf(
            "dad@refrigerationservices.com.au",
            "jane.doe@refrigerationservices.com.au",
            "service+jobs@refrigerationservices.com.au",
            "DAD@REFRIGERATIONSERVICES.COM.AU",
            "  spaced@refrigerationservices.com.au  ",
        ).forEach { assertTrue("expected $it to be allowed", Recipients.isAllowed(it)) }
    }

    @Test
    fun `rejects everything else including lookalikes`() {
        listOf(
            "someone@gmail.com",
            "someone@refrigerationservices.com",
            "someone@refrigerationservices.com.au.evil.com",
            "someone@sub.refrigerationservices.com.au",
            "someone@notrefrigerationservices.com.au",
            "a@refrigerationservices.com.au, b@evil.com",
            "a@refrigerationservices.com.au;b@evil.com",
            "\"a@refrigerationservices.com.au\" <b@evil.com>",
            "@refrigerationservices.com.au",
            "refrigerationservices.com.au",
            "",
            null,
        ).forEach { assertFalse("expected $it to be rejected", Recipients.isAllowed(it)) }
    }

    @Test
    fun `header injection attempts are rejected`() {
        assertFalse(Recipients.isAllowed("a@refrigerationservices.com.au\nbcc: b@evil.com"))
    }

    @Test
    fun `uses the exact wording the business asked for`() {
        val result = Recipients.check("stranger@gmail.com")
        assertTrue(result is Recipients.Result.Invalid)
        assertEquals(
            "This email is not one of your family member",
            (result as Recipients.Result.Invalid).error,
        )
    }

    @Test
    fun `blank input is a different problem to a wrong domain`() {
        val result = Recipients.check("   ")
        assertTrue(result is Recipients.Result.Invalid)
        assertEquals("Enter an email address first.", (result as Recipients.Result.Invalid).error)
    }

    @Test
    fun `accepted addresses are normalised`() {
        val result = Recipients.check("  Dad@Refrigerationservices.Com.Au ")
        assertTrue(result is Recipients.Result.Valid)
        assertEquals("dad@refrigerationservices.com.au", (result as Recipients.Result.Valid).email)
    }
}
